packages/gatsby/src/redux/remove-nodes-by-plugin.js

```diff
@@ -0,0 +1,58 @@
+const isIterable = obj =>
+  obj !== null && typeof obj[Symbol.iterator] === `function`
+
+const getLookup = state => {
+  let lookup = new Map()
+  for (let [typeOrId, nodeOrNodes] of state) {
+    if (isIterable(nodeOrNodes)) {
+      for (let [id, node] of nodeOrNodes) {
+        lookup.set(id, node)
+      }
+    } else {
+      lookup.set(typeOrId, nodeOrNodes)
+    }
+  }
+  return lookup
+}
+
+const getDirtyNodes = (plugins, state) => {
+  let dirty = new Map()
+  for (let [id, node] of state) {
+    if (node && node.internal && plugins.includes(node.internal.owner)) {
+      dirty.set(id, true)
+      let parent = state[node.parent]
+      while (parent) {
+        dirty.set(parent.id, true)
+        parent = state[parent.parent]
+      }
+    }
+  }
+  return dirty
+}
+
+/*
+ * This invalidates nodes (in one loop!) and their parent node(s)
+ * that match a plugin name (e.g. node.internal.owner)
+ * when a plugin version changes
+ */
+module.exports = (plugin, state, storeByType = true) => {
```

Comment:
This is a slightly janky utility to:

1. Augment the redux state for nodes to remove nodes (and parent nodes) that have been "touched" from the previous run
1. It will receive (e.g. `['gatsby-source-filesystem', 'gatsby-source-remote-images']`) as an argument, and then will invalidate any nodes owned by specified plugin(s) and their parent nodes
    - This _works_ but needs some TLC and some more time to validate the approach